src/Servers/Kestrel/Core/src/Internal/Http/Http1UpgradeMessageBody.cs

```diff
@@ -69,12 +72,49 @@ public override Task StopAsync()
 
         public override bool TryReadInternal(out ReadResult readResult)
         {
-            return _context.Input.TryRead(out readResult);
+            // Ignore the canceled readResult unless it was canceled by the user.
+            do
+            {
+                if (!_context.Input.TryRead(out readResult))
```

Comment:
Could this end up in a tight loop until cancellation is consumed by the main loop?

Reply:
Yes, if the PipeReader repeatedly returns a canceled ReadResult. Same goes ReadAsync. It should still be safe, because PipeReaders are supposed to only return at most 1 canceled ReadResult per call to CancelPendingRead(). And this logic isn't entirely new, even in 3.1. [Http1ContentLengthMessageBody](https://github.com/dotnet/aspnetcore/blob/84e3ff5b38841e88765cde2fe6f6267f1b3cc8cd/src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs#L141-L156) already has this behavior.

One difference you'll notice though between this and Http1ContentLengthMessageBody is that it doesn't call AdvanceTo before reading again after a canceled ReadResult (Remember #25799). This isn't necessary for the DefaultPipeReader and simplifies the logic quite a bit, but I could easily seeing this breaking custom ones. That said, if it breaks because of this, it would wind up just throwing an InvalidOperationException instead of the OperationCanceledException it throws today.

I understand being nervous though. Given severity of the issue (which is basically throwing from WebSockets 5 seconds earlier than we should during shutdown), do you still think this is worth backporting to 3.1?

Reply:
Yes we should still back port it.